### DISCUSSION TITLE
Is there any practical need for multiple "multi" handles?

### DISCUSSION THREAD

**Discussion by determin1st** [curl/curl#11741] [Q&A]
Im abstracting curl api in php and the question about having multiple multi handles arouse. 
```php
$curler_1 = new CurlBasedFetcher($options);
$curler_2 = new CurlBasedFetcher($anotherOptions);
```
should these two instances reuse the same multi handle loop or separation is better?
multi handle i create have basically the same options (`CURLMOPT_PIPELINING=1`)

in other words, does anything outside of `CURLMOPT_*` create the need for multiple multi handles?

**Comment by bagder** [comment]
> in other words, does anything outside of CURLMOPT_* create the need for multiple multi handles?

Maybe if you want to run them in separate threads?

**Reply by determin1st** [reply]
im running them in the same, single php process, but they can be different baseUrl or different hostnames in `CURLOPT_URL` so, i assumed that multiplexing wont work when there are different hosts in the same multi-handle or maybe there are some other option that would downgrade multi-handle group

**Reply by bagder** [reply]
Multiplexing can't work unless they're the same host name, but I can't see any reason to use two multi handles for this. It seems to only make things more complicated.

**Reply by determin1st** [reply]
do i understand it correctly that this all-in-one:
```
multi-handle [
  easy-handle-1: https://example.com/getSomeData
  easy-handle-2: https://example.com/postInfo
  easy-handle-3: https://another.us/id/12345
  easy-handle-4: https://another.us/id/67890
]
```
will neither do multiplexing for example.com, nor for another.us, but this:
```
multi-handle-1 [
  easy-handle-1: https://example.com/getSomeData
  easy-handle-2: https://example.com/postInfo
]
multi-handle-2 [
  easy-handle-1: https://another.us/id/12345
  easy-handle-2: https://another.us/id/67890
]
```
will do both?

**Reply by bagder** [reply]
I'm not sure I understand the question. curl (if asked) will do multiplexing when it can, but the transfers (easy handles) of course need to be added to the same multi handle.

**Reply by determin1st** [reply]
the question is whether multiplexing would still work for certain hostname/site when it is not unique among easy-handles in the given multi-handle:
```
multi-handle [
  easy-handle-1: https://example.com/getSomeData
  easy-handle-2: https://example.com/postInfo
  easy-handle-3: https://another.us/id/12345
  easy-handle-4: https://another.us/id/67890
]
```
would the `example.com` and `another.us` transfers become multiplexed? (knowing that both servers support it)

i only tested multiplexing for a single hostname, that is api.telegram.org, but asynchronous application may deal with multiple resources which may support multiplexing, so its better to know beforehead whether each instance of simplified curl fetcher (http-only protocol) requires its own multi-handle for such a feature.

**Reply by bagder** [reply]
curl (if asked) will do multiplexing when it can. If you have many more connections around that it can't use for multiplexing does not change that.
